core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala

```diff
@@ -71,6 +82,161 @@ private[spark] class CoarseGrainedExecutorBackend(
+        val userExecUnitsConfigName =
+          SPARK_EXECUTOR_RESOURCE_PREFIX + rName + SPARK_RESOURCE_UNITS_POSTFIX
+        val userExecConfigUnits = env.conf.getOption(userExecUnitsConfigName)
+        val realCounts = if (execResourceInfo.units.nonEmpty) {
+          if (taskUnits.nonEmpty && taskUnits.get.nonEmpty) {
+            if (userExecConfigUnits.isEmpty || userExecConfigUnits.get.isEmpty) {
+              throw new SparkException(s"Resource: $rName has units in task config " +
+                s"and executor startup config but the user specified executor resource " +
+                s"config is missing the units config - see ${userExecUnitsConfigName}.")
+            }
+            try {
+              val execCountWithUnits =
+                Utils.byteStringAsBytes(execResourceInfo.count.toString + execResourceInfo.units)
+              val taskCountWithUnits = Utils.byteStringAsBytes(taskCount + taskUnits.get)
+              ResourceRealCounts(execCountWithUnits, taskCountWithUnits)
+            } catch {
+              case e: NumberFormatException =>
+                // Ignore units not of byte types and just use count
+                logWarning(s"Illegal resource unit type, spark only " +
+                  s"supports conversion of byte types, units: $execResourceInfo.units, " +
+                  s"ignoring the type and using the raw count.", e)
+                ResourceRealCounts(execResourceInfo.count, taskCount.toLong)
+            }
+          } else {
+            throw new SparkException(
+              s"Resource: $rName has an executor units config: ${execResourceInfo.units}, but " +
+                s"the task units config is missing.")
+          }
+        } else {
+          if (taskUnits.nonEmpty && taskUnits.get.nonEmpty) {
+            throw new SparkException(
+              s"Resource: $rName has a task units config: ${taskUnits.get}, but the executor " +
+              s"units config is missing.")
+          }
+          ResourceRealCounts(execResourceInfo.count, taskCount.toLong)
+        }
+        if (realCounts.execCount < realCounts.taskCount) {
+          throw new SparkException(s"Executor resource: $rName, count: ${realCounts.execCount} " +
+            s"isn't large enough to meet task requirements of: ${realCounts.taskCount}")
+        }
+        // also make sure the executor resource count on start matches the
+        // spark.executor.resource configs specified by user
+        val userExecCountConfigName =
+          SPARK_EXECUTOR_RESOURCE_PREFIX + rName + SPARK_RESOURCE_COUNT_POSTFIX
+        val userExecConfigCount = env.conf.getOption(userExecCountConfigName).
+          getOrElse(throw new SparkException(s"Executor resource: $rName not specified " +
+            s"via config: $userExecCountConfigName, but required " +
+            s"by the task, please fix your configuration"))
+        val execConfigCountWithUnits = if (userExecConfigUnits.nonEmpty) {
+          val count = try {
```

Comment:
nit: could we combine the similar code with L117 ~L127 ?

Reply:
Sure